api/provider/aws/aws_test/load_balancer_create_test.go

```diff
@@ -0,0 +1,335 @@
+				Return(listServerCertificatesOutput, nil)
+
+			listener.SetSSLCertificateId(port.CertificateName)
+		}
+
+		switch strings.ToLower(port.Protocol) {
+		case "http", "https":
+			listener.SetInstanceProtocol("http")
+		case "tcp", "ssl":
+			listener.SetInstanceProtocol("tcp")
+		}
+
+		listeners[i] = listener
+	}
+
+	createLoadBalancerInput := &elb.CreateLoadBalancerInput{}
+	createLoadBalancerInput.SetLoadBalancerName("l0-test-lb_id")
+	createLoadBalancerInput.SetScheme("internal")
+	createLoadBalancerInput.SetSecurityGroups([]*string{aws.String("env_sg"), aws.String("lb_sg")})
+	createLoadBalancerInput.SetSubnets([]*string{aws.String("priv1"), aws.String("priv2")})
+	createLoadBalancerInput.SetListeners(listeners)
+
+	mockAWS.ELB.EXPECT().
+		CreateLoadBalancer(createLoadBalancerInput).
+		Return(&elb.CreateLoadBalancerOutput{}, nil)
+
+	healthCheck := &elb.HealthCheck{}
+	healthCheck.SetTarget(req.HealthCheck.Target)
+	healthCheck.SetInterval(int64(req.HealthCheck.Interval))
+	healthCheck.SetTimeout(int64(req.HealthCheck.Timeout))
+	healthCheck.SetHealthyThreshold(int64(req.HealthCheck.HealthyThreshold))
+	healthCheck.SetUnhealthyThreshold(int64(req.HealthCheck.UnhealthyThreshold))
+
+	configureHealthCheckInput := &elb.ConfigureHealthCheckInput{}
+	configureHealthCheckInput.SetLoadBalancerName("l0-test-lb_id")
+	configureHealthCheckInput.SetHealthCheck(healthCheck)
+
+	mockAWS.ELB.EXPECT().
+		ConfigureHealthCheck(configureHealthCheckInput).
+		Return(&elb.ConfigureHealthCheckOutput{}, nil)
+
+	target := provider.NewLoadBalancerProvider(mockAWS.Client(), tagStore, mockConfig)
+	result, err := target.Create(req)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	assert.Equal(t, "lb_id", result)
+
+	expectedTags := models.Tags{
+		{
+			EntityID:   "lb_id",
+			EntityType: "load_balancer",
+			Key:        "name",
+			Value:      "lb_name",
+		},
+		{
+			EntityID:   "lb_id",
+			EntityType: "load_balancer",
+			Key:        "environment_id",
+			Value:      "env_id",
+		},
+	}
+
+	for _, tag := range expectedTags {
+		assert.Contains(t, tagStore.Tags(), tag)
+	}
+}
+
+func TestLoadBalancerCreateDefaults(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockAWS := awsc.NewMockClient(ctrl)
+	tagStore := tag.NewMemoryStore()
+	mockConfig := mock_config.NewMockAPIConfig(ctrl)
+
+	mockConfig.EXPECT().Instance().Return("test").AnyTimes()
+	mockConfig.EXPECT().VPC().Return("vpc_id").AnyTimes()
+	mockConfig.EXPECT().Region().Return("region").AnyTimes()
+	mockConfig.EXPECT().AccountID().Return("123456789012")
+	mockConfig.EXPECT().PrivateSubnets().Return([]string{"priv1", "priv2"}).AnyTimes()
+	mockConfig.EXPECT().PublicSubnets().Return([]string{"pub1", "pub2"}).AnyTimes()
+
+	defer provider.SetEntityIDGenerator("lb_id")()
+
+	req := models.CreateLoadBalancerRequest{
+		LoadBalancerName: "lb_name",
+		EnvironmentID:    "env_id",
+		IsPublic:         true,
+		Ports: []models.Port{
+			{
+				ContainerPort: 80,
```

Comment:
we should have at least 2 ports, and we shouldn't test with `ContainerPort` and `HostPort` with the same value. They should be different so we verify they map correctly to loadbalancer/instance ports

Reply:
This is the `TestLoadBalancerCreateDefaults` test we're looking at here. Should I be defining 2 ports in a default checking unit test such as this?

Reply:
To your other point, something I'm also immediately concerned about is the implementation of the entity method action `Create()`. The reason why I use `port.HostPort` [for both the From and To ports](https://github.com/quintilesims/layer0/blob/a256730c903ef3983edfa569c55403e9284e50a1/api/provider/aws/aws_test/load_balancer_create_test.go#L73) is because of the helper function [authorizeSGIngressFromPort](https://github.com/quintilesims/layer0/blob/api-refactor/api/provider/aws/load_balancer_create.go#L155) which takes in the load balancer security group ID and a single `int64` port.

When [this helper function is called](https://github.com/quintilesims/layer0/blob/api-refactor/api/provider/aws/load_balancer_create.go#L62), we are passing in the iterated on `req.Ports.HostPort`. 

The thing that would case this to break would be `input.SetToPort` in the helper function `authorizeSGIngressFromPort`, because this clearly would need to be the `req.Ports.ContainerPort`.

Can I conclude that this is a bug in the helper function then?

Reply:
Actually, disregard my last comment. I'm conflating the port mappings and the ingress security group. The port mappings should be different, but the security group ingress authorization for the load balancer will always use the host port for To and From because traffic from the public will go through the load balancer and to the host port. The container will forward traffic from the host to the listening container port.

So, this is by design.

Still, given that we are looking at the Defaults unit test, I'm curious if you think I still need to define 2 ports for that test.

Reply:
Sorry, meant for the 2 ports to be in the 'master' test. I assume a default of 0 ports is allowed?